torchrl/data/utils.py

```diff
@@ -35,6 +43,206 @@
+            ):
+                lazy_key_example = lazy_keys_examples[lazy_key]
+                sub_td.set(
+                    lazy_key,
+                    _empty_like(lazy_key_example, sub_td.batch_size),
+                )
+            td.tensordicts[td_index] = sub_td
+
+    if recurse_through_entries:
+        for key in td.keys():
+            try:
+                td.set(
+                    key,
+                    unlazyfy_keys(
+                        td.get(key), recurse_through_entries, recurse_through_stack
+                    ),
+                )
+            except RuntimeError as err:
+                if re.match(
+                    r"Found more than one unique shape in the tensors", str(err)
+                ):
+                    pass  # no need to unlazify since key is het leaf
+                else:
+                    raise err
+
+    return td
+
+
+def relazyfy_keys(
+    td,
+):
+    """Given a TensorDictBase, restores lazy keys by removing 0 shaped tensors and related orphan tensordicts."""
+    if not is_tensor_collection(td):
+        return None if td.numel() == 0 else td.clone()
+
+    td = td.clone()
+
+    if isinstance(td, LazyStackedTensorDict):
+        for td_index in range(len(td.tensordicts)):
+            sub_td = td.tensordicts[td_index]
+            sub_td = relazyfy_keys(sub_td)
+            td.tensordicts[td_index] = sub_td
+
+    for key in list(td.keys()):
+        try:
+            value = relazyfy_keys(td.get(key))
+            if value is None:
+                del td[key]
+            else:
+                td.set(
+                    key,
+                    value,
+                )
+        except RuntimeError as err:
+            if re.match(r"Found more than one unique shape in the tensors", str(err)):
+                pass
+            else:
+                raise err
+
+    if isinstance(td, TensorDict) and not len(td.keys()):
+        return None
+    return td
+
+
+def _empty_like(td, batch_size):
+    if is_tensor_collection(td):
+        return td.empty()
+    else:
+        shape = [dim if i < len(batch_size) else 0 for i, dim in enumerate(td.shape)]
+
+        return torch.empty(
+            shape,
+            dtype=td.dtype,
+            device=td.device,
+        )
+
+
+def check_no_lazy_keys(td, recurse: bool = True):
+    """Given a TensorDictBase, returns true if there are no lazy keys."""
+    if isinstance(td, LazyStackedTensorDict):
+        keys = set(td.keys())
+        for t in td.tensordicts:
+            if recurse and not check_no_lazy_keys(t):
+                return False
+            if set(t.keys()) != keys:
+                return False
+    elif isinstance(td, TensorDict) and recurse:
+        for i in td.values():
+            if not check_no_lazy_keys(i):
+                return False
+    elif isinstance(td, torch.Tensor):
+        return True
+    else:
+        return False
+
+    return True
+
+
+def get_all_keys(td, include_lazy: bool):
```

Comment:
is this something that we will keep in torchrl? My feeling is that it should not be exposed to the users

Reply:
i didn't put it in tensordict because of the planned deprecation of lazy keys.
it was here cause i thought we might want to use it in a few places, but now i have moved it to the tests